### ISSUE TITLE
Optimizer and LRScheduler step() calls seem to be in wrong order

### ISSUE THREAD

**Issue by joaqo** [JDAI-CV/fast-reid#403]
Hi, I am running a training job using the the current default [VERI-Wild config](https://github.com/JDAI-CV/fast-reid/blob/master/configs/VERIWild/bagtricks_R50-ibn.yml) and I am getting the following warning:

```UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate```

At the same time I am getting `Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 32768.0` warnings at the begginning of training, so it seems I am starting with a LR that is way too large when using the default configuration. I didn't have any luck finding a way to invert the order of the `step()` calls, as the codebase is quite complex.

Anyone know how can I fix this? Thanks!

**Comment by L1aoXingyu** [comment]
It is not a bug, gradient overflow is due to fp16 mixed-precision training. You can check this in nvidia/apex.

**Comment by workingcoder** [comment]
UserWarning about the order of lr_scheduler.step() and optimizer.step()：
This is because there is a learning rate adjustment in the warmup stage in the training process, which is placed after each iter, while optimizer.step() is placed after each epoch. The warmup stage causes the warning.
During the stage after warmup, lr_scheduler.step() is placed after optimizer.step() according to the right order of pytorch suggestion.
So you can ignore this warning, the program is running correctly.

**Comment by L1aoXingyu** [comment]
Actually, `optimizer.step()` is placed after each iter, so this warning is caused by fp16 training.

When you turn off `fp16`, this warning will not appear. When `fp16` is enabled, there may cause gradient overflow problem, so it will skip the `optimizer.step()` in this iter. Then this warning will appear.

**Comment by workingcoder** [comment]
> Actually, `optimizer.step()` is placed after each iter, so this warning is caused by fp16 training.
> 
> When you turn off `fp16`, this warning will not appear. When `fp16` is enabled, there may cause gradient overflow problem, so it will skip the `optimizer.step()` in this iter. Then this warning will appear.

Well, I got it wrong. Thank you for your reply. (^-^)

**Comment by joaqo** [comment]
Thanks for the answer! Just to check, the message ```UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`...``` is caused by fp16 mixed-precision training? Is this a pytorch bug in their warning messages? Because the warning message is completely wrong then!
